File: src/libqtipccommon.cpp
#include <QNativeIpcKey>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qtipccommon.h>
#include "libqtipccommon.h"
#include "libqtipccommon.hxx"

QNativeIpcKey* QNativeIpcKey_new() {
    return new QNativeIpcKey();
}

QNativeIpcKey* QNativeIpcKey_new2(uint16_t typeVal) {
    return new QNativeIpcKey(static_cast<QNativeIpcKey::Type>(typeVal));
}

QNativeIpcKey* QNativeIpcKey_new3(const libqt_string k) {
    QString k_QString = QString::fromUtf8(k.data, k.len);
    return new QNativeIpcKey(k_QString);
}

QNativeIpcKey* QNativeIpcKey_new4(const QNativeIpcKey* other) {
    return new QNativeIpcKey(*other);
}

QNativeIpcKey* QNativeIpcKey_new5(const libqt_string k, uint16_t typeVal) {
    QString k_QString = QString::fromUtf8(k.data, k.len);
    return new QNativeIpcKey(k_QString, static_cast<QNativeIpcKey::Type>(typeVal));
}

uint16_t QNativeIpcKey_LegacyDefaultTypeForOs() {
    return static_cast<uint16_t>(QNativeIpcKey::legacyDefaultTypeForOs());
}

void QNativeIpcKey_OperatorAssign(QNativeIpcKey* self, const QNativeIpcKey* other) {
    self->operator=(*other);
}

void QNativeIpcKey_Swap(QNativeIpcKey* self, QNativeIpcKey* other) {
    self->swap(*other);
}

bool QNativeIpcKey_IsEmpty(const QNativeIpcKey* self) {
    return self->isEmpty();
}

bool QNativeIpcKey_IsValid(const QNativeIpcKey* self) {
    return self->isValid();
}

uint16_t QNativeIpcKey_Type(const QNativeIpcKey* self) {
    return static_cast<uint16_t>(self->type());
}

void QNativeIpcKey_SetType(QNativeIpcKey* self, uint16_t typeVal) {
    self->setType(static_cast<QNativeIpcKey::Type>(typeVal));
}

libqt_string QNativeIpcKey_NativeKey(const QNativeIpcKey* self) {
    QString _ret = self->nativeKey();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QNativeIpcKey_SetNativeKey(QNativeIpcKey* self, const libqt_string newKey) {
    QString newKey_QString = QString::fromUtf8(newKey.data, newKey.len);
    self->setNativeKey(newKey_QString);
}

libqt_string QNativeIpcKey_ToString(const QNativeIpcKey* self) {
    QString _ret = self->toString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QNativeIpcKey* QNativeIpcKey_FromString(const libqt_string stringVal) {
    QString stringVal_QString = QString::fromUtf8(stringVal.data, stringVal.len);
    return new QNativeIpcKey(QNativeIpcKey::fromString(stringVal_QString));
}

void QNativeIpcKey_Delete(QNativeIpcKey* self) {
    delete self;
}

File: include/libqtipccommon.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qtipccommon_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qnativeipckey.html
pub const qnativeipckey = struct {
    /// New constructs a new QNativeIpcKey object.
    ///
    ///
    pub fn New() QtC.QNativeIpcKey {
        return qtc.QNativeIpcKey_new();
    }

    /// New2 constructs a new QNativeIpcKey object.
    ///
    /// ``` typeVal: qtipccommon_enums.Type ```
    pub fn New2(typeVal: i64) QtC.QNativeIpcKey {
        return qtc.QNativeIpcKey_new2(@intCast(typeVal));
    }

    /// New3 constructs a new QNativeIpcKey object.
    ///
    /// ``` k: []const u8 ```
    pub fn New3(k: []const u8) QtC.QNativeIpcKey {
        const k_str = qtc.struct_libqt_string{
            .len = k.len,
            .data = k.ptr,
        };

        return qtc.QNativeIpcKey_new3(k_str);
    }

    /// New4 constructs a new QNativeIpcKey object.
    ///
    /// ``` other: QtC.QNativeIpcKey ```
    pub fn New4(other: ?*anyopaque) QtC.QNativeIpcKey {
        return qtc.QNativeIpcKey_new4(@ptrCast(other));
    }

    /// New5 constructs a new QNativeIpcKey object.
    ///
    /// ``` k: []const u8, typeVal: qtipccommon_enums.Type ```
    pub fn New5(k: []const u8, typeVal: i64) QtC.QNativeIpcKey {
        const k_str = qtc.struct_libqt_string{
            .len = k.len,
            .data = k.ptr,
        };

        return qtc.QNativeIpcKey_new5(k_str, @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#legacyDefaultTypeForOs)
    ///
    ///
    pub fn LegacyDefaultTypeForOs() i64 {
        return qtc.QNativeIpcKey_LegacyDefaultTypeForOs();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#operator-eq)
    ///
    /// ``` self: QtC.QNativeIpcKey, other: QtC.QNativeIpcKey ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QNativeIpcKey_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#swap)
    ///
    /// ``` self: QtC.QNativeIpcKey, other: QtC.QNativeIpcKey ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QNativeIpcKey_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#isEmpty)
    ///
    /// ``` self: QtC.QNativeIpcKey ```
    pub fn IsEmpty(self: ?*anyopaque) bool {
        return qtc.QNativeIpcKey_IsEmpty(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#isValid)
    ///
    /// ``` self: QtC.QNativeIpcKey ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QNativeIpcKey_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#type)
    ///
    /// ``` self: QtC.QNativeIpcKey ```
    pub fn Type(self: ?*anyopaque) i64 {
        return qtc.QNativeIpcKey_Type(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#setType)
    ///
    /// ``` self: QtC.QNativeIpcKey, typeVal: qtipccommon_enums.Type ```
    pub fn SetType(self: ?*anyopaque, typeVal: i64) void {
        qtc.QNativeIpcKey_SetType(@ptrCast(self), @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#nativeKey)
    ///
    /// ``` self: QtC.QNativeIpcKey, allocator: std.mem.Allocator ```
    pub fn NativeKey(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QNativeIpcKey_NativeKey(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qnativeipckey.NativeKey: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#setNativeKey)
    ///
    /// ``` self: QtC.QNativeIpcKey, newKey: []const u8 ```
    pub fn SetNativeKey(self: ?*anyopaque, newKey: []const u8) void {
        const newKey_str = qtc.struct_libqt_string{
            .len = newKey.len,
            .data = newKey.ptr,
        };
        qtc.QNativeIpcKey_SetNativeKey(@ptrCast(self), newKey_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#toString)
    ///
    /// ``` self: QtC.QNativeIpcKey, allocator: std.mem.Allocator ```
    pub fn ToString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QNativeIpcKey_ToString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qnativeipckey.ToString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#fromString)
    ///
    /// ``` stringVal: []const u8 ```
    pub fn FromString(stringVal: []const u8) QtC.QNativeIpcKey {
        const stringVal_str = qtc.struct_libqt_string{
            .len = stringVal.len,
            .data = stringVal.ptr,
        };
        return qtc.QNativeIpcKey_FromString(stringVal_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qnativeipckey.html#dtor.QNativeIpcKey)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QNativeIpcKey ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QNativeIpcKey_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qtipccommon.html#types
pub const enums = struct {
    pub const Type = enum {
        pub const SystemV: u16 = 81;
        pub const PosixRealtime: u16 = 256;
        pub const Windows: u16 = 257;
    };
};
